### PR TITLE
Feature/jr2 852 receive and transform from avam

### PR DESCRIPTION
**Pull request by michael-wirth**

### CODE REVIEW

File: infrastructure-messagebroker-shared/src/main/java/ch/admin/seco/jobs/services/jobadservice/infrastructure/messagebroker/messages/ApproveJobAdvertisementMessage.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+import ch.admin.seco.jobs.services.jobadservice.domain.jobadvertisement.Occupation;
+
+public class ApproveJobAdvertisementMessage {
+
+    private String stellennummerEgov;
+
+    private boolean reportingObligation;
+
+    private LocalDate reportingObligationEndDate;
```
Comment (cjpsycos) [inline infrastructure-messagebroker-shared/src/main/java/ch/admin/seco/jobs/services/jobadservice/infrastructure/messagebroker/messages/ApproveJobAdvertisementMessage.java]: missing approvalDate

File: infrastructure-messagebroker-shared/src/main/java/ch/admin/seco/jobs/services/jobadservice/infrastructure/messagebroker/messages/CreateOrUpdateJobAdvertisementMessage.java [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+
+    private Company company;
+
+    private Contact contact;
+
+    private Location location;
+
+    private List<Occupation> occupations;
+
+    private List<LanguageSkill> languageSkills;
```
Comment (cjpsycos) [inline infrastructure-messagebroker-shared/src/main/java/ch/admin/seco/jobs/services/jobadservice/infrastructure/messagebroker/messages/CreateOrUpdateJobAdvertisementMessage.java]: many fields missing

File: infrastructure-messagebroker-shared/src/main/java/ch/admin/seco/jobs/services/jobadservice/infrastructure/messagebroker/messages/RejectJobAdvertisementMessage.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+import java.time.LocalDate;
+
+public class RejectJobAdvertisementMessage {
+
+    private String stellennummerEgov;
+
+    private LocalDate rejectionDate;
+
+    private String rejectionCode;
```
Comment (cjpsycos) [inline infrastructure-messagebroker-shared/src/main/java/ch/admin/seco/jobs/services/jobadservice/infrastructure/messagebroker/messages/RejectJobAdvertisementMessage.java]: missing rejectionReason

File: infrastructure-messagebroker-shared/src/main/java/ch/admin/seco/jobs/services/jobadservice/infrastructure/messagebroker/messages/RejectJobAdvertisementMessage.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
+package ch.admin.seco.jobs.services.jobadservice.infrastructure.messagebroker.messages;
+
+import java.time.LocalDate;
+
+public class RejectJobAdvertisementMessage {
+
+    private String stellennummerEgov;
```
Comment (cjpsycos) [inline infrastructure-messagebroker-shared/src/main/java/ch/admin/seco/jobs/services/jobadservice/infrastructure/messagebroker/messages/RejectJobAdvertisementMessage.java]: missing stellennummerAvam

File: infrastructure-messagebroker-shared/src/main/java/ch/admin/seco/jobs/services/jobadservice/infrastructure/messagebroker/messages/CancelJobAdvertisementMessage.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
+package ch.admin.seco.jobs.services.jobadservice.infrastructure.messagebroker.messages;
+
+import java.time.LocalDate;
+
+public class CancelJobAdvertisementMessage {
+
+    private String stellennummerAvam;
```
Comment (cjpsycos) [inline infrastructure-messagebroker-shared/src/main/java/ch/admin/seco/jobs/services/jobadservice/infrastructure/messagebroker/messages/CancelJobAdvertisementMessage.java]: missing stellennummerEgov

File: infrastructure-messagebroker-shared/src/main/java/ch/admin/seco/jobs/services/jobadservice/infrastructure/messagebroker/messages/ApproveJobAdvertisementMessage.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
+package ch.admin.seco.jobs.services.jobadservice.infrastructure.messagebroker.messages;
+
+import java.time.LocalDate;
+import java.util.List;
+
+import ch.admin.seco.jobs.services.jobadservice.domain.jobadvertisement.Occupation;
+
+public class ApproveJobAdvertisementMessage {
+
+    private String stellennummerEgov;
```
Comment (cjpsycos) [inline infrastructure-messagebroker-shared/src/main/java/ch/admin/seco/jobs/services/jobadservice/infrastructure/messagebroker/messages/ApproveJobAdvertisementMessage.java]: missing stellennummerAvam

File: infrastructure-messagebroker-shared/src/main/java/ch/admin/seco/jobs/services/jobadservice/infrastructure/messagebroker/messages/UpdateJobAdvertisementMessage.java [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+import ch.admin.seco.jobs.services.jobadservice.domain.jobadvertisement.Contact;
+import ch.admin.seco.jobs.services.jobadservice.domain.jobadvertisement.Employment;
+import ch.admin.seco.jobs.services.jobadservice.domain.jobadvertisement.LanguageSkill;
+import ch.admin.seco.jobs.services.jobadservice.domain.jobadvertisement.Location;
+import ch.admin.seco.jobs.services.jobadservice.domain.jobadvertisement.Occupation;
+import ch.admin.seco.jobs.services.jobadservice.domain.jobadvertisement.SourceSystem;
+
+public class UpdateJobAdvertisementMessage {
+
+    private String stellennummerAvam;
```
Comment (cjpsycos) [inline infrastructure-messagebroker-shared/src/main/java/ch/admin/seco/jobs/services/jobadservice/infrastructure/messagebroker/messages/UpdateJobAdvertisementMessage.java]: missing stellennummerEgov

File: infrastructure-messagebroker-ws-avam-source/src/main/java/ch/admin/seco/jobs/services/jobadservice/infrastructure/messagebroker/avam/AvamEndpoint.java [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+import org.springframework.ws.server.endpoint.annotation.RequestPayload;
+import org.springframework.ws.server.endpoint.annotation.ResponsePayload;
+
+import ch.admin.seco.jobs.services.jobadservice.infrastructure.ws.avam.source.InsertOste;
+import ch.admin.seco.jobs.services.jobadservice.infrastructure.ws.avam.source.InsertOsteResponse;
+import ch.admin.seco.jobs.services.jobadservice.infrastructure.ws.avam.source.WSOsteEgov;
+
+@Endpoint
+public class AvamEndpoint {
+    private static final Logger LOGGER = LoggerFactory.getLogger(AvamEndpoint.class);
```
Comment (cjpsycos) [inline infrastructure-messagebroker-ws-avam-source/src/main/java/ch/admin/seco/jobs/services/jobadservice/infrastructure/messagebroker/avam/AvamEndpoint.java]: We should name it the same way everywhere: LOGGER, LOG, log, ...
Maybe separated task for cleanup this

File: infrastructure-messagebroker-ws-avam-source/src/main/java/ch/admin/seco/jobs/services/jobadservice/infrastructure/messagebroker/avam/AvamEndpoint.java [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+
+    private static final String SECO_WS_OK = "SECO_WS: OK";
+    private static final String SECO_WS_ERROR = "SECO_WS: ERROR";
```
Comment (cjpsycos) [inline infrastructure-messagebroker-ws-avam-source/src/main/java/ch/admin/seco/jobs/services/jobadservice/infrastructure/messagebroker/avam/AvamEndpoint.java]: rename to eg. RESPONSE_OK / RESPONSE_ERROR so I see it is a response code/term

File: infrastructure-messagebroker-ws-avam-source/src/main/java/ch/admin/seco/jobs/services/jobadservice/infrastructure/messagebroker/avam/AvamEndpoint.java [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+                avamSource.reject(jobAdvertisementFromAvamAssembler.createRejectJobAdvertisement(oste));
+            } else if (isApproved(oste)) {
+                avamSource.approve(jobAdvertisementFromAvamAssembler.createApproveJobAdvertisement(oste));
+            } else if (isCreatedOrUpdatedByRAV(oste)) {
+                avamSource.createOrUpdate(jobAdvertisementFromAvamAssembler.createUpdateJobAdvertisement(oste));
+            } else if (isCanceledByRAV(oste)) {
+                avamSource.cancel(jobAdvertisementFromAvamAssembler.createCancelJobAdvertisement(oste));
+            } else {
+                LOGGER.warn("Received JobAdvertisement in unknown state from AVAM: {}", transformToXml(request));
+                return createReponse(SECO_WS_ERROR);
```
Comment (cjpsycos) [inline infrastructure-messagebroker-ws-avam-source/src/main/java/ch/admin/seco/jobs/services/jobadservice/infrastructure/messagebroker/avam/AvamEndpoint.java]: Why are we creating a response?

File: infrastructure-messagebroker-ws-avam-source/src/main/java/ch/admin/seco/jobs/services/jobadservice/infrastructure/messagebroker/avam/AvamEndpoint.java [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+
+        } catch (JAXBException e) {
+            LOGGER.error("Marshalling of JaxbObject failed", e);
+            return xmlRootObject.toString();
+        }
+    }
+
+    private boolean isCreatedOrUpdatedByRAV(WSOsteEgov oste) {
+        return isFromRAV(oste) && !isCanceled(oste);
+
```
Comment (cjpsycos) [inline infrastructure-messagebroker-ws-avam-source/src/main/java/ch/admin/seco/jobs/services/jobadservice/infrastructure/messagebroker/avam/AvamEndpoint.java]: empty line

File: infrastructure-messagebroker-ws-avam-source/src/main/java/ch/admin/seco/jobs/services/jobadservice/infrastructure/messagebroker/avam/AvamEndpoint.java (hunk 8 above)
Comment (birom) [reply infrastructure-messagebroker-ws-avam-source/src/main/java/ch/admin/seco/jobs/services/jobadservice/infrastructure/messagebroker/avam/AvamEndpoint.java]: And sometimes it is an instance variable and sometimes static.

File: infrastructure-messagebroker-ws-avam-source/src/main/java/ch/admin/seco/jobs/services/jobadservice/infrastructure/messagebroker/avam/AvamEndpoint.java (hunk 9 above)
Comment (varusakov) [reply infrastructure-messagebroker-ws-avam-source/src/main/java/ch/admin/seco/jobs/services/jobadservice/infrastructure/messagebroker/avam/AvamEndpoint.java]: @cjpsycos should it be `RESPONSE_OK = "SECO_WS: OK"`?

File: infrastructure-messagebroker-shared/src/main/java/ch/admin/seco/jobs/services/jobadservice/infrastructure/messagebroker/messages/CreateOrUpdateJobAdvertisementMessage.java (hunk 2 above)
Comment (varusakov) [reply infrastructure-messagebroker-shared/src/main/java/ch/admin/seco/jobs/services/jobadservice/infrastructure/messagebroker/messages/CreateOrUpdateJobAdvertisementMessage.java]: @cjpsycos where I can check all required fields?

File: infrastructure-messagebroker-ws-avam-source/src/main/java/ch/admin/seco/jobs/services/jobadservice/infrastructure/messagebroker/avam/AvamEndpoint.java (hunk 9 above)
Comment (cjpsycos) [reply infrastructure-messagebroker-ws-avam-source/src/main/java/ch/admin/seco/jobs/services/jobadservice/infrastructure/messagebroker/avam/AvamEndpoint.java]: yes

File: domain/src/main/java/ch/admin/seco/jobs/services/jobadservice/domain/jobadvertisement/JobAdvertisementFactory.java [hunk 12]
```diff
@@ -57,17 +59,20 @@ public JobAdvertisement createFromApi(Locale language, String title, String desc
...
-        JobAdvertisement jobAdvertisement = new JobAdvertisement.Builder()
-                .setId(new JobAdvertisementId())
-                .setSourceSystem(SourceSystem.RAV)
-                .setStatus(JobAdvertisementStatus.REFINING)
-                .setLanguage(language)
-                .setTitle(title)
-                .setDescription(description)
-                .setReportToRav(true)
-                .build();
+        JobAdvertisement jobAdvertisement = jobAdvertisementRepository.findByStellennummerAvam(stellennummerAvam)
```
Comment (cjpsycos) [inline domain/src/main/java/ch/admin/seco/jobs/services/jobadservice/domain/jobadvertisement/JobAdvertisementFactory.java]: This is a factory for creation and NOT for updating existing data. The application service could have a method receiveFromAvam, and there decide to create or update

File: .travis.yml [hunk 13]
```diff
@@ -31,7 +31,8 @@ deploy:
   script: "./gradlew bootJar publish -Pprod -Pzipkin -x test  -Pversion=${TRAVIS_JOB_NUMBER}-SNAPSHOT --stacktrace
     && curl https://dev.job-room.ch:8443/deploy/jr2 --insecure"
   on:
-    branch: master
+    branch: develop
```
Comment (michael-wirth) [inline .travis.yml]: should be `master`

File: application/src/main/java/ch/admin/seco/jobs/services/jobadservice/application/jobadvertisement/JobAdvertisementApplicationService.java [hunk 14]
```diff
@@ -156,6 +157,67 @@ public JobAdvertisementId createFromApi(CreateJobAdvertisementApiDto createJobAd
         return jobAdvertisement.getId();
     }
 
+    public JobAdvertisementId createFromAvam(CreateJobAdvertisementAvamDto createJobAdvertisementAvamDto) {
+        LOG.debug("Create or update '{}' from AVAM", createJobAdvertisementAvamDto.getTitle());
```
Comment (michael-wirth) [inline application/src/main/java/ch/admin/seco/jobs/services/jobadservice/application/jobadvertisement/JobAdvertisementApplicationService.java]: Change log message to just `Create '{}' from AVAM`

File: application/src/main/java/ch/admin/seco/jobs/services/jobadservice/application/jobadvertisement/JobAdvertisementApplicationService.java [hunk 15]
```diff
@@ -156,6 +157,67 @@ public JobAdvertisementId createFromApi(CreateJobAdvertisementApiDto createJobAd
...
+                .setEures(createJobAdvertisementAvamDto.getPublication().isEures())
+                .setEuresAnonymous(createJobAdvertisementAvamDto.getPublication().isEuresAnonymous())
+                .setPublicDisplay(createJobAdvertisementAvamDto.getPublication().isPublicDisplay())
+                .setPublicAnonynomous(createJobAdvertisementAvamDto.getPublication().isPublicAnonynomous())
+                .setRestrictedDisplay(createJobAdvertisementAvamDto.getPublication().isRestrictedDisplay())
+                .setRestrictedAnonymous(createJobAdvertisementAvamDto.getPublication().isRestrictedAnonymous())
+                .build();
+
+        final JobAdvertisementCreator creator = new JobAdvertisementCreator.Builder(null)
+                // FIXME .setReportingObligation(createJobAdvertisementAvamDto.getReportingObligation())
```
Comment (michael-wirth) [inline application/src/main/java/ch/admin/seco/jobs/services/jobadservice/application/jobadvertisement/JobAdvertisementApplicationService.java]: What is the issue here? Do we need any other data than the new WSDL?

File: infrastructure-messagebroker-ws-avam-source/src/main/java/ch/admin/seco/jobs/services/jobadservice/infrastructure/messagebroker/avam/v1/AvamEndpointV1.java [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+import org.springframework.ws.server.endpoint.annotation.PayloadRoot;
+import org.springframework.ws.server.endpoint.annotation.RequestPayload;
+import org.springframework.ws.server.endpoint.annotation.ResponsePayload;
+
+import ch.admin.seco.jobs.services.jobadservice.infrastructure.ws.avam.source.v1.InsertOste;
+import ch.admin.seco.jobs.services.jobadservice.infrastructure.ws.avam.source.v1.InsertOsteResponse;
+import ch.admin.seco.jobs.services.jobadservice.infrastructure.ws.avam.source.v1.WSOsteEgov;
+
+@Endpoint
+public class AvamEndpointV1 {
```
Comment (michael-wirth) [inline infrastructure-messagebroker-ws-avam-source/src/main/java/ch/admin/seco/jobs/services/jobadservice/infrastructure/messagebroker/avam/v1/AvamEndpointV1.java]: Do we need Profiles to start V1 or V2?

**Pull request merged**